[ECL] Moving towards multi-sensor-per-same-measurement fusion

kamilritz opened the pull request:
This ECL [PR](https://github.com/PX4/ecl/pull/652) aims to provide unambiguous logging of velocity and position measurement updates in the EKF2 estimator. 

To enable this, we need to provide the right uorb messages on the Firmware side. Making the measurements sensor specific would increase the number of fields in `ekf2_innovations.msg`. 
To avoid this we split the message in to sensor specific innovations, innovation variances and innovation test ratios.
These messages have different use-cases and can be enabled accordingly.

Logs of several test flights can be found in the ECL PR.

The `ekf2_innovations.msg` is marked as deprecated. The fields are replaced according to the following scheme: 

ekf2_innovations.vel_pos_innov			-> 		estimator_innovations.[gps_hvel, gps_vvel, gps_hpos, gps_vpos, ev_hvel, ev_vvel, ev_hpos, ev_vpos, fake_hvel, fake_vvel, fake_hpos, fake_vpos ]
ekf2_innovations.mag_innov				-> 		estimator_innovations.mag
ekf2_innovations.heading_innov			-> 		estimator_innovations.heading
ekf2_innovations.airspeed_innov			-> 		estimator_innovations.airspeed
ekf2_innovations.beta_innov				-> 		estimator_innovations.beta
ekf2_innovations.flow_innov				-> 		estimator_innovations.flow
ekf2_innovations.hagl_innov				-> 		estimator_innovations.hagl
ekf2_innovations.vel_pos_innov_var		-> 		estimator_innovation_variances.[gps_hvel, gps_vvel, gps_hpos, gps_vpos, ev_hvel, ev_vvel, ev_hpos, ev_vpos, fake_hvel, fake_vvel, fake_hpos, fake_vpos ]
ekf2_innovations.mag_innov_var			-> 		estimator_innovation_variances.mag
ekf2_innovations.heading_innov_var		-> 		estimator_innovation_variances.heading
ekf2_innovations.airspeed_innov_var		-> 		estimator_innovation_variances.airspeed
ekf2_innovations.beta_innov_var			-> 		estimator_innovation_variances.beta
ekf2_innovations.flow_innov_var			-> 		estimator_innovation_variances.flow
ekf2_innovations.hagl_innov_var			-> 		estimator_innovation_variances.hagl
ekf2_innovations.output_tracking_error	-> 		estimator_status.output_tracking_error
ekf2_innovations.drag_innov				-> 		estimator_innovations.drag
ekf2_innovations.drag_innov_var			-> 		estimator_innovation_variances.drag
ekf2_innovations.aux_vel_innov			-> 		estimator_innovations.[aux_hvel, aux_vvel]
not existing							->		estimator_innovation_variances.[aux_hvel, aux_vvel]

kamilritz commented:
@dagar This is running out of space on fmu-v2 :)
Do you have an idea how we can get this in?

bresch commented:
Looks good

TSC21 commented:
@kamilritz you are also missing adding the RTPS ID's: http://ci.px4.io:8080/blue/organizations/jenkins/PX4_misc%2FFirmware-compile/detail/PR-13127/8/pipeline/60#step-619-log-45

TSC21 commented:
> @kamilritz is there any reason ware creating 2 new messages for this? Why not just use `ekf2_innovations.msg` and add `#TOPICS` aliases for the other two, as they seem to have the exact same fields?

This would be something like: (`ekf2_innovation.msg`)
```
uint64 timestamp	# time since system start (microseconds)

# GPS
float32 gps_hvel	# horizontal GPS velocity innovation
float32 gps_vvel	# vertical GPS velocity innovation
float32 gps_hpos	# horizontal GPS position innovation
float32 gps_vpos	# vertical GPS position innovation

# External Vision
float32 ev_hvel	# horizontal external vision velocity innovation
float32 ev_vvel	# vertical external vision velocity innovation
float32 ev_hpos	# horizontal external vision position innovation
float32 ev_vpos	# vertical external vision position innovation

# Height sensors
float32 rng_vpos	# range sensor height innovation
float32 baro_vpos	# barometer height innovation

# Auxiliary (horizontal) velocity
float32 aux_hvel	# horizontal auxiliar velocity innovation from landing target measurement
float32 flow		# flow innvoation test ratio

# Various
float32 heading	# heading innovation
float32 mag		# earth magnetic field innovation
float32[2] drag	# drag specific force innovation
float32 airspeed	# airspeed innovation
float32 beta		# synthetic sideslip innovation
float32 hagl		# height of ground innovation

#TOPICS ekf2_innovations ekf2_innovation_variances ekf2_innovation_test_ratios
```

Also, please add those aliases to the RTPS ID's list.

kamilritz commented:
@TSC21 That is a nice idea. The innovation_test_ratios fields are not exactly the same. But we can do this for innovation and innovation variances.

TSC21 commented:
> @TSC21 That is a nice idea. The innovation_test_ratios fields are not exactly the same. But we can do this for innovation and innovation variances.

@kamilritz the only difference I see is on the arrays. The same structure can still be used, but instead of using the entire array, just used the first array index.

TSC21 commented on msg/ekf2_innovations.msg:
```diff
@@ -1,19 +1,34 @@
+
+# TOPICS ekf2_innovations ekf2_innovation_variances ekf2_innovation_test_ratios
+
+# For the innovation test ratios are scalar values. In case the field is a vector,
```
Please bring this above the `#TOPICS` line (let the `#TOPICS` line be the last one.

TSC21 commented on msg/ekf2_innovations.msg:
```diff
@@ -1,19 +1,34 @@
+float32[2] flow		# flow innvoation
+
+# Various
+float32 heading		# heading innovation
```
Can you make the comment indentation more uniform?

kamilritz replied:
I think the indentation is okay. Have a look at the file itself.

TSC21 replied:
Sometimes the IDEs trick us ;) I do trust the Github indentation

kamilritz replied:
Yeah, I know. Use the "view file" option on Github. Seems that Github tricks us as well ;)

TSC21 commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -284,6 +284,8 @@ class Ekf2 final : public ModuleBase<Ekf2>, public ModuleParams, public px4::Wor
 	vehicle_status_s		_vehicle_status{};
 
 	uORB::Publication<ekf2_innovations_s>			_estimator_innovations_pub{ORB_ID(ekf2_innovations)};
+	uORB::Publication<ekf2_innovations_s>				_estimator_innovation_variances_pub{ORB_ID(ekf2_innovation_variances)};
```
Indentation

TSC21 commented on src/modules/local_position_estimator/sensors/mocap.cpp:
```diff
@@ -144,15 +144,20 @@ void BlockLocalPositionEstimator::mocapCorrect()
-		_pub_innov.get().vel_pos_innov[i] = 0;
-		_pub_innov.get().vel_pos_innov_var[i] = 1;
-	}
+	_pub_innov.get().ev_hvel[0] = 0.0f;
```
Should this ben NAN instead?

kamilritz replied:
Yeah, that would make sense.

TSC21 commented on src/modules/local_position_estimator/sensors/vision.cpp:
```diff
@@ -161,15 +161,20 @@ void BlockLocalPositionEstimator::visionCorrect()
-		_pub_innov.get().vel_pos_innov[i] = 0;
-		_pub_innov.get().vel_pos_innov_var[i] = 1;
-	}
+	_pub_innov.get().ev_hvel[0] = 0.0f;
```
NAN?

TSC21 commented on src/modules/local_position_estimator/sensors/vision.cpp:
```diff
@@ -161,15 +161,20 @@ void BlockLocalPositionEstimator::visionCorrect()
+	_pub_innov.get().ev_vpos    = r(2, 0);
+
+	// publish innovation variances
+	_pub_innov_var.get().ev_hvel[0] = 1.0f;
```
NAN?

TSC21 commented:
> @dagar This is running out of space on fmu-v2 :)
> Do you have an idea how we can get this in?

I am not sure how much short we can get on fmu-v2 to get this in :S

TSC21 commented on msg/tools/uorb_rtps_message_ids.yaml:
```diff
@@ -325,4 +325,10 @@ rtps:
   - msg: vehicle_visual_odometry_aligned
     id: 169
     alias: vehicle_odometry
+    msg: ekf2_innovation_variances
```
you are missing the dash here:
```suggestion
  - msg: ekf2_innovation_variances
```

TSC21 commented on msg/tools/uorb_rtps_message_ids.yaml:
```diff
@@ -325,4 +325,10 @@ rtps:
+    msg: ekf2_innovation_variances
+    id: 170
+    alias: ekf2_innovations
+    msg: ekf2_innovation_test_ratios
```
```suggestion
  - msg: ekf2_innovation_test_ratios
```

dagar commented:
> > @dagar This is running out of space on fmu-v2 :)
> > Do you have an idea how we can get this in?
> 
> I am not sure how much short we can get on fmu-v2 to get this in :S

There's quite a bit more that can be saved with a little effort. https://github.com/PX4/Firmware/pull/12601 https://github.com/PX4/Firmware/pull/12191

TSC21 commented:
> > > @dagar This is running out of space on fmu-v2 :)
> > > Do you have an idea how we can get this in?
> > 
> > 
> > I am not sure how much short we can get on fmu-v2 to get this in :S
> 
> There's quite a bit more that can be saved with a little effort. #12601 #12191

Still on time for a release?

LorenzMeier commented:
@TSC21 @kamilritz We can't change messages in ulog wildly like that. There are two options:

1) Just extend the existing message, only adding fields and not renaming any
2) Rename the whole message, so that the post-flight tooling can check for different message formats.

TSC21 commented:
> @TSC21 @kamilritz We can't change messages in ulog wildly like that. There are two options:
> 
> 1. Just extend the existing message, only adding fields and not renaming any
> 2. Rename the whole message, so that the post-flight tooling can check for different message formats.

@LorenzMeier I was with the idea that it was ok to rename fields as long as we were synced with a release period were the debugging and plotting tools could be updated as well - and in that case it were the plotting tools that were extended and not the uORB msgs with new fields. But if it's just too much field renaming, I guess we can just add new fields and keep the previous naming for the others, but I think we lose consistency here.

jkflying commented:
> This would break all post-flight tooling due to the uorb format changes - we need to take a more compliant approach for message changes. It is fine to break uORB to some degree, but ulog needs to be a more steady format.

Let's make an RFC for this and add it to the guidelines then, also go through what we think are critical messages and add a SITL test to make sure they still work.

CarlOlsson commented:
> Let's make an RFC for this and add it to the guidelines then, also go through what we think are critical messages and add a SITL test to make sure they still work for log processing tools.

Shouldn't log processing tools be topic name/field agnostic? And if they are not they can map the names from different releases to a common name?

kamilritz commented:
I changed the topic names such that it is detectable, which version of the message is logged.

jkflying commented:
> Shouldn't log processing tools be topic name/field agnostic? And if they are not they can map the names from different releases to a common name?

I think the issue is something like https://logs.px4.io where it automatically puts certain topics on certain graphs. If we reuse a field without changing the name these tools will silently start plotting garbage.

mhkabir commented on msg/estimator_status.msg:
```diff
@@ -9,6 +9,8 @@ float32[3] vibe			# IMU vibration metrics in the following array locations
 
 float32[24] covariances	# Diagonal Elements of Covariance Matrix
 
+float32[3] output_tracking_error # return a vector containing the output predictor angular, velocity and position tracking error magnitudes (rad), (m/s), (m)
```
Can you split this up into 3 self-documenting fields? Having this as a vector makes things opaque.

kamilritz replied:
Agree that this would be clearer. These message changes affect some log analytics in the CI. I prefer to make it not more complicated for now.

mhkabir replied:
I don't get it. This is a new field you're adding in this PR. How is this already used in / affect analytics?

kamilritz replied:
I moved it from the ekf2_innovations message to the the estimator_status message

mhkabir replied:
Can you address my original comment?

kamilritz replied:
The output_tracking_error is a vector that is constructed in ECL EKF. It contains different metrics similar to the estimator_status.vibe field, and the different components are described well in the comments. Since we would need to change things in a lot of different places, I would prefer leaving it like this.

mhkabir commented on msg/ekf2_innovations.msg:
```diff
@@ -1,19 +1 @@
-float32[2] drag_innov		# drag specific force innovation (m/s/s)
-float32[2] drag_innov_var	# drag specific force innovation variance (m/s/s)**2
-float32[2] aux_vel_innov	# auxiliary NE velocity innovations from landing target measurement (m/s)
+# This message is DEPRECATED by the estimator_innovations message.
```
Don't think we need to keep this file.

kamilritz replied:
@TSC21 is working on a better solution. Having a deprecated message list.

mhkabir commented on msg/estimator_innovations.msg:
```diff
@@ -0,0 +1,41 @@
+# Auxiliary velocity
+float32[2] aux_hvel	# horizontal auxiliar velocity innovation from landing target measurement (m/sec) and innovation variance ((m/sec)**2)
+float32    aux_vvel	# vertical auxiliar velocity innovation from landing target measurement (m/sec) and innovation variance ((m/sec)**2)
+float32[2] flow		# flow innvoation (rad/sec) and innovation variance ((rad/sec)**2)
```
Put this under its own comment, instead of auxvel?

kamilritz replied:
Going to add this.

mhkabir commented on msg/estimator_innovations.msg:
```diff
@@ -0,0 +1,41 @@
+float32[2] flow		# flow innvoation (rad/sec) and innovation variance ((rad/sec)**2)
+
+# Various
+float32 heading		# heading innovation (rad) and innovation variance (rad**2)
```
What if I have multiple heading sources? Compass, GPS Yaw and VIO are some options.

kamilritz replied:
That is a good observation. That can be tackled when someone is working on multi sensor heading fusion.

mhkabir replied:
The thing is, this message is not easily changeable in the future, and every change is a breaking one due to the log parsing tools.

kamilritz replied:
One solution could be adding now fields such that we would have [gps_heading, ev_heading, mag_heading, heading] and only use the heading field. And after adding the future changes still fill the heading field in addition to all the other, but mark the heading field as deprecated. This would give log users time to adopt. What do you think about that approach?

kamilritz replied:
I would prefer adding the additional and at the moment unused heading field at the moment when we at the functionality.

mhkabir commented on src/modules/local_position_estimator/sensors/flow.cpp:
```diff
@@ -172,10 +172,10 @@ void BlockLocalPositionEstimator::flowCorrect()
+	_pub_innov.get().flow[0] = r(0);
+	_pub_innov_var.get().flow[1] = r(1);
+	_pub_innov.get().flow[0] = S(0, 0);
+	_pub_innov_var.get().flow[1] = S(1, 1);
```
This is incorrect, please check.

kamilritz replied:
Thank you very much for catching that. I really appreciate your detailed reviews.

mhkabir commented on src/systemcmds/topic_listener/generate_listener.py:
```diff
@@ -24,6 +24,15 @@
 for index,m in enumerate(raw_messages):
 	topic_list = []
 	f = open(m,'r')
+	has_timestamp = False
```
A timestamp should be required in all messages. What is the rationale here?

kamilritz replied:
The idea was to leave deprecated uorb message defintions, as an empty file and remove it from the CMakeLists. In order for the listener_generator not to add empty, deprecated message definition, this check was added. This will be obsolete after adding the deprecated message list.

mhkabir replied:
What is the goal of leaving it as an empty file? I feel like it doesn't add any value and we can just remove it.

mhkabir replied:
This change can be removed now.

mhkabir commented on src/modules/ekf2/Utility/PreFlightChecker.cpp:
```diff
@@ -92,16 +93,16 @@ bool PreFlightChecker::preFlightCheckHorizVelFailed(const ekf2_innovations_s &in
+bool PreFlightChecker::preFlightCheckVertVelFailed(const estimator_innovations_s &innov, const float alpha)
 {
-	const float vel_d_innov = innov.vel_pos_innov[2];
+	const float vel_d_innov = fmaxf(fabsf(innov.gps_vvel), fabs(innov.ev_vvel));
```
This seems a bit sketchy. What if you have terrible GPS vel, but good vision vel? Should this be checking for the minimum instead?

kamilritz replied:
I also was thinking about this. Same applies for velocity and position innovation checks during missions in Commander.

kamilritz replied:
The issue is that innovations are published as zero if there is no fusion happening. So we would need to reject the zero case. For the moment, taking the max is the conservative approach.

mhkabir replied:
Not convinced that this is the correct approach. It seems to me that this will just always fail if enabled in a scenario where GPS is available but bad. Have you done any testing with the vertical velocity preflight check enabled in such a scenario? If yes, can you share the log please?

kamilritz replied:
Isn't the vertical velocity preflight check enabled by default? If not, how would I enable it?

jkflying replied:
IMO: longer term the preflight checks shouldn't use innovations at all , but rather the covariances of the states. In the mean time this is a conservative check that can be performed, but I think we all agree it isn't optimal.

mhkabir commented on src/modules/ekf2/Utility/PreFlightChecker.cpp:
```diff
@@ -92,16 +93,16 @@ bool PreFlightChecker::preFlightCheckHorizVelFailed(const ekf2_innovations_s &in
+bool PreFlightChecker::preFlightCheckHeightFailed(const estimator_innovations_s &innov, const float alpha)
 {
-	const float hgt_innov = innov.vel_pos_innov[5];
+	const float hgt_innov = innov.gps_vpos;
```
Why GPS only?

kamilritz replied:
Currently, the height fusion is using only one sensor at a time. ECL is publishing them on the gps_vpos. I will add a comment to clarify this.

kamilritz commented:
@dagar Can we get this in? I think I have sufficiently addressed the feedback and CI is only failing due to limited stack size on fmu v2.
If yes, we first need to merge: https://github.com/PX4/ecl/pull/668

kamilritz commented:
@mhkabir I have tested it on my platform extensively. I ran GPS, VIO, FLOW in a lot of combinations and never had a preflight check failure that was not magnetometer related :)

mhkabir commented:
@kamilritz Can you post some of that test data?

@PX4/testflights can you test this please?

mhkabir commented on src/modules/ekf2/Utility/PreFlightChecker.cpp:
```diff
@@ -92,16 +93,16 @@ bool PreFlightChecker::preFlightCheckHorizVelFailed(const ekf2_innovations_s &in
+bool PreFlightChecker::preFlightCheckHeightFailed(const estimator_innovations_s &innov, const float alpha)
 {
-	const float hgt_innov = innov.vel_pos_innov[5];
+	const float hgt_innov = innov.gps_vpos;    // height innovation independent of sensor is published on gps_vpos
```
It should be noted that this is a hack.

mhkabir commented:
@jkflying @bkueng Can you review this? This is the biggest change to the EKF (and the ekf2 frontend) in a while.

kamilritz commented:
@mhkabir This is a log with GPS, VIO and FLOW combined: https://review.px4.io/plot_app?log=47bfecee-e43a-465c-82aa-d48f796e9749

mhkabir commented:
@kamilritz thanks for the log. I looks OK, apart from this:

![image](https://user-images.githubusercontent.com/5557676/69573272-c2bd0c80-0f93-11ea-95b0-eaf11f10a0e2.png)

kamilritz commented:
> @kamilritz thanks for the log. I looks OK, apart from this:

Is there something I can do about this?

dannyfpv commented:
tested  on pixhawk4 v5 f-450
Flight Card 1

Modes Tested:
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.

Procedure:
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

Notes:
No issues were noted, good flight in general.

Logs:
https://review.px4.io/plot_app?log=a671f8b5-6372-47d1-97dc-e4b4d0a3a379Flight Card 2

Modes Tested
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

Procedure

Armed form QGC.
Took-off on mission plan mode
. Note that commands were sent form QGC (Armed and takeoff)
Once all waypoint completed vehicle switched to RTL and landed as expected.
Good flight in general.
Note:
No issues were noted, good flight in general.

Logs:  https://review.px4.io/plot_app?log=38bf258c-29c1-43c2-ba2e-2f1df34e5af4

Flight Card 3
Modes Tested
Position Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.
Good flight in general.

Note:
No issues were noted, good flight in general.

Logs:   https://review.px4.io/plot_app?log=a2b30f1d-1a01-40a6-b305-69ca7692d8cf

Flight Card 4
Modes Tested
Fail safe:The radio control turned off and the vehicle returned and landed correctly.
Tested good.

Logs:
https://review.px4.io/plot_app?log=523e05cc-2d63-460a-8792-518502e1efc1

Junkim3DR commented:
### Tested on NXP FMUK66 v3

**Flight Card 1**
- Modes Tested
  - Position Mode: Good
  - Altitude Mode: Good
  -Stabilized Mode: Good

- Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

- Notes
No issues noted, good flight in general.

- Log
https://review.px4.io/plot_app?log=815b10b1-6ff9-413a-9d37-4e36f4df7598

**Flight Card 2**
- Modes Tested
  -Mission Plan Mode (Automated): Good.
  -RTL (Return To Land): Good.

- Procedure
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

- Notes
Vehicle did not land properly.
While landing, vehicle descended very quickly then stoped at about 3 meters from the ground and ,like a bounce, the vehicle increased altitude to about 10 meters then it proceeded to rush to the ground fast enough for it to bounce off the ground and a really violent way and flipped. (Pilot required to switch "Kill Switch" ON in order to avoid vehicle damaging motors and propellers after the bounce/flip).
Loaded Master and vehicle behaved as expected, no issues.

- Log
https://review.px4.io/plot_app?log=1a70df8e-aeac-47c3-87ac-ad39b73b8e00

**Flight Card 3**
- Modes Tested
  - Position Mode: Good.
  - Mission Plan Mode (Automated): Good.
  - RTL (Return To Land): Good.

- Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

- Notes
Vehicle did not land properly.
While landing, vehicle descended very quickly then stoped at about 3 meters from the ground and ,like a bounce, the vehicle increased altitude to about 10 meters then it proceeded to rush to the ground fast enough for it to bounce off the ground and a really violent way and flipped. (Pilot required to switch "Kill Switch" ON in order to avoid vehicle damaging motors and propellers after the bounce/flip). 
Loaded Master and vehicle behaved as expected, no issues.

- Log
https://review.px4.io/plot_app?log=7ae7fbcb-9531-47d1-9267-dd0bd772e4f4






### Tested on Pixhawk Pro v4

**Flight Card 1**
- Modes Tested
  - Position Mode: Good
  - Altitude Mode: Good
  -Stabilized Mode: Good

- Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

- Notes
No issues noted, good flight in general.

- Log
https://review.px4.io/plot_app?log=f1a26e95-b979-467d-83d1-5dc0c5fb4fdd

**Flight Card 2**
- Modes Tested
  -Mission Plan Mode (Automated): Good.
  -RTL (Return To Land): Good.

- Procedure
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

- Notes
No issues noted, good flight in general.

- Log
https://review.px4.io/plot_app?log=ac657dc9-733b-471b-907c-cd0bab7f772a

**Flight Card 3**
- Modes Tested
  - Position Mode: Good.
  - Mission Plan Mode (Automated): Good.
  - RTL (Return To Land): Good.

- Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

- Notes
No issues noted, good flight in general.

- Log
https://review.px4.io/plot_app?log=aab917c9-aa48-42be-a439-b857f7266e1a

**Flight Card 4**
- Modes Tested
  - Position Mode: Good.
  - Failsafe RTL (Return To Land): Not returning after RC shutdown.

- Procedure
Arm and Take off in position mode, after flying for approximately one minute, turned RC off in order to activate failsafe RTL, wait for vehicle to return and land as expected.

- Notes
No issue noted, good flight in general.

- Log
https://review.px4.io/plot_app?log=c0a1bc66-9c00-43f3-a03e-431b6bae2613

jkflying commented on src/modules/local_position_estimator/sensors/gps.cpp:
```diff
@@ -189,10 +189,20 @@ void BlockLocalPositionEstimator::gpsCorrect()
-		_pub_innov.get().vel_pos_innov[i] = r(i);
-		_pub_innov.get().vel_pos_innov_var[i] = S(i, i);
-	}
+	_pub_innov.get().gps_hvel[0] = r(3);
```
Reorder these so that in `r` and `S` it goes in ascending order.

kamilritz replied:
Done

jkflying commented on src/modules/local_position_estimator/sensors/mocap.cpp:
```diff
@@ -144,15 +144,20 @@ void BlockLocalPositionEstimator::mocapCorrect()
-		_pub_innov.get().vel_pos_innov_var[i] = S(i, i);
-	}
-
-	for (size_t i = 3; i < 6; i++) {
```
Just as a note, this was broken before

MaEtUgR commented:
> Loaded Master and vehicle behaved as expected, no issues.

@Junkim3DR To see if the reported problems in landing detection are for sure related to this pr and what exactly was different on that particular setup it would help a lot to have the log when you compared with master and had no issues as well. Do you still have this log? Could you upload it?

Junkim3DR commented:
> Do you still have this log? Could you upload it?

Hi @MaEtUgR, I believe this are the 2 logs I tested with master to see if the issue persisted.

**Logs**
  - https://review.px4.io/plot_app?log=2c48c8eb-b7b5-4322-840a-be6b7d75886a
  - https://review.px4.io/plot_app?log=eb0718a7-b018-400d-8a1b-1ae1fe71c5b6

Tony3dr commented:
@MaEtUgR it's the same setup nothing was changed.

kamilritz commented:
@Junkim3DR can you please fly the NXP FMUK66 v3 with flight card 2&3 again. Once with this PR and once with current master. Please perform the same flights with both firmware version. Thank you very much.

mhkabir commented:
Remaining things before we merge this (we're very close!):
1. FMU-v2 flash overflow.
2. Merge ECL PR. and rebase this on 
3. Make sure land detection failure is not related to changes here.

kamilritz commented:
> 4. Make sure land detection failure is not related to changes here.

@mhkabir any ideas how I can investigate this?

Junkim3DR commented:
### Tested on NXP FMUK66 v3

**Flight Card 2**
- Modes Tested
  -Mission Plan Mode (Automated): Good.
  -RTL (Return To Land): Good.

- Procedure
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

- Notes
No issues on PR, vehicle landed slightly faster on master and bounced from the ground considerably.

- Log
  - PR: https://review.px4.io/plot_app?log=358d8b6c-1e76-483b-b42f-f549f5ffe7bf
  - Master: https://review.px4.io/plot_app?log=c2c7dfaa-dd65-4e85-b8ad-835cc90efe58


**Flight Card 3**
- Modes Tested
  - Position Mode: Good.
  - Mission Plan Mode (Automated): Good.
  - RTL (Return To Land): Good.

- Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

- Notes
Vehicle had no issue at all.

- Log
  - PR: https://review.px4.io/plot_app?log=f1a87df3-f4bd-4960-bed5-12f24ec68249
  - Master: https://review.px4.io/plot_app?log=9123ef2f-c716-4925-b899-f04965c53b16

kamilritz commented:
@Junkim3DR Thanks for the new log files. There is definitively something odd. We would need another log file that we can use for EKF replays. Can you flash the newest version of this PR, set SDLOG_PROFILE=2, set SDLOG_MODE="from boot to disarm", fly Flight Card 2 again and upload the log.
Thank you very much.

kamilritz commented:
@Junkim3DR Which master firmware were you running as reference?

Junkim3DR commented:
> We would need another log file that we can use for EKF replays. Can you flash the newest version of this PR, set SDLOG_PROFILE=2, set SDLOG_MODE="from boot to disarm", fly Flight Card 2 again and upload the log.

### Tested on NXP FMUK66 v3

**Flight Card 2**
- Modes Tested
  -Mission Plan Mode (Automated): Good.
  -RTL (Return To Land): Good.

- Procedure
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

- Notes
No issues on PR, vehicle landed slightly faster on master and bounced from the ground considerably.

- Log
  - PR13127 1st Flight: https://review.px4.io/plot_app?log=72bcbcb2-0a47-487b-9ee5-b51c9ebc30ee       
  - PR13127 2nd Flight: https://review.px4.io/plot_app?log=71dba018-52b6-4e8d-ad7f-00a867a5bdf0




> Which master firmware were you running as reference?

Master used was latest form QGC the 12th of December 2019, commit: 9b66cbd2

kamilritz commented:
@Junkim3DR Thank you very much!

@mhkabir @dagar  Together with @priseborough we tracked down the issue in the first test flights to extensively filtered baro data on the `vehicle_air_data`. On my test platform and in the newest log files on the NXP FMUK66 platform the baro data seems to be noisy as expected and the weird behavior during landing caused by bad vertical estimates is not noticeable.

I argue that it is okay to merge this now. LGTM!!!! (Reminder: ECL PR has to go first: https://github.com/PX4/ecl/pull/668)

priseborough commented:
I've checked the logs from Mexico testers who were using the standard GPS sensor fit and daa looked OK.
